From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Giuliano Procida <gprocida@google.com>
Date: Wed, 28 Oct 2020 12:35:54 +0000
Subject: Stabilise sort of canonical types

Some types like unnamed-enum-underlying-type are not distinguished by
type_topo_comp. This can result in nondeterministic output and flakey
tests.

While a more complete ordering from type_topo_comp would be nice, the
nondeterminism can reduced by preserving the relative order of
identically-named types.

	* src/abg-ir.cc (scope_decl::get_sorted_canonical_types): Sort
	canonical types with std::stable_sort(..., type_topo_comp()).

Signed-off-by: Giuliano Procida <gprocida@google.com>
Reviewed-by: Matthias Maennich <maennich@google.com>
Signed-off-by: Matthias Maennich <maennich@google.com>
---
 src/abg-ir.cc | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/abg-ir.cc b/src/abg-ir.cc
index a9323aad0013..6c1dc9d88c4e 100644
--- a/src/abg-ir.cc
+++ b/src/abg-ir.cc
@@ -6108,9 +6108,9 @@ scope_decl::get_sorted_canonical_types() const
 	priv_->sorted_canonical_types_.push_back(*e);
 
       type_topo_comp comp;
-      std::sort(priv_->sorted_canonical_types_.begin(),
-		priv_->sorted_canonical_types_.end(),
-		comp);
+      std::stable_sort(priv_->sorted_canonical_types_.begin(),
+		       priv_->sorted_canonical_types_.end(),
+		       comp);
     }
   return priv_->sorted_canonical_types_;
 }
